### DISCUSSION TITLE
How to keep the header and body of http request in same tcp segment in .net 4.5?

### DISCUSSION THREAD

**Discussion by yll690** [dotnet/runtime#74304] [Q&A] [answered]
Hi, everyone. I ran into a http issue when I try to post some data to a server using .net 4.5. I can post successfully via Postman but in .net I always get 502 error. I assume it is the server's problem but I can do nothing to the server. 

The I found the main difference between the requests sended by Postman and .net 4.5 is that **the body is in a seperated segment in .net.** Then I tried .net 6.0, it behaves like Postman and can post  successfully. However I can't migrate my project to .net 6.0 now. So is there any way to control the segmentation in .net 4.5?

This is the test code both in .net 4.5 and .net 6.0:
```
    HttpClient client = new HttpClient();
    HttpContent content = new ByteArrayContent(Encoding.UTF8.GetBytes("asdasd"));
    HttpResponseMessage response = client.PostAsync("http://192.168.110.150:10002/api/sdk/message", content).Result;
```

What the tcp server received:
![Snipaste_2022-08-20_16-53-40](https://user-images.githubusercontent.com/16819902/185737729-9df8b87d-1b35-48a5-99f5-f645b112a066.png)

Note that I tested those code in C# Interactive which should use 4.8 in latest windows 10 and it makes no difference.

**Comment by huoyaoyuan** [comment]
@dotnet/ncl 

Note that .NET Framework 4.5 should be out of support now. Using 4.8 which is an in-place update is suggested.

**Reply by yll690** [reply]
Yes, but I tested those code in C# Interactive which should use 4.8 in latest windows 10 and it makes no difference.

**Comment by rzikm** [comment]
I don't think you can influence this in any way. This seems to be a property of how `WinHttpHandler` works (or the underlying native lib, didn't investigate further). You can reproduce the same behavior on .NET 6 if you explicitly create `HttpClient` with `WinHttpHandler`.

.NET Core by default uses managed implementation of Http (`SocketsHttpHandler`) which does stuff a little differently which leads to the body being sent in the same TCP packet as the headers.

**Reply by rzikm** [reply]
Actually, .NET framework uses WebRequest under the hood in HttpClient, but the point stands, I don't think there is a way to force putting both header and body in the same TCP segment

**Comment by MihaZupan** [comment, marked as answer]
This is really fragile behavior for the server to rely on.
If you can't upgrade to .NET 6, you can consider putting a reverse proxy between your client and server. The client will still send the request in multiple TCP segments, but the proxy can buffer them for you.

With [YARP](https://microsoft.github.io/reverse-proxy/) for example, you would get the same behavior as if you had upgraded the client to 6.0.
We will send up to 4 KB of headers+content down to the `Socket` in 1 call, but there is no guarantee of how that will be split up on the transport layer. If your headers+content are always small, odds are they will be grouped together.

**Reply by yll690** [reply]
Yes, using a proxy seems the best way. Finally I write a simple proxy tool using .net 6.0 which solves the problem. Thank you.

**Comment by wfurt** [comment]
As mentioned above, all the options are fragile and there is no good way how to impact this reliably. You can also try to set `ServicePointManager.Expect100Continue = true` (or set the header explicitly) With that, compliant server would send you intermediate response before expecting request data. You can also check/set `ServicePointManager.UseNagleAlgorithm` That would increase chance of writing small chunks of data together. 
However the server implementation seems broken so you may be out of luck

**Reply by yll690** [reply]
Yes, it mainly the server's problem. `ServicePointManager` does not help. Thank you anyway.

**Comment by yll690** [comment]
Thank you very much. Fanally I didn't find any way to influence the segmentation. So I write a simple proxy tool using .net 6.0 which solves the problem.
